### PR TITLE
Feature/trade blotter

### PR DESCRIPTION
**Pull request by laura-b2c2**
This still needs some tests added and styling of the table improved as it's quite difficult to read but hopefully all the data is correct.

Using React Window to virtualise the grid. We have to be pretty precise with dimensions for this otherwise there are some scroll issues (i.e. it scrolls the nav out of view or the body scrolls out of line with the headers) just something we need to be aware of if changing cell widths or adding other components to the page etc. 

Currently looks like:
image

### CODE REVIEW

File: src/components/trades/TradesList.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import AutoSizer from 'react-virtualized-auto-sizer';
+
+const TABLE_WIDTH = 1365;
+
+const StyledRow = styled.div`
+  display: flex;
+  width: ${TABLE_WIDTH}px
+`
+
+const Row = ({ style, data, index }: ListChildComponentProps<Array<TradeData>>) => {
```
Comment (jim-b2c2) [inline src/components/trades/TradesList.tsx:36]: `ListChildComponentProps<TradeData[]>` I think you can just do - I would stick with literals as oppose to constructors (unless we are actually using them)

File: src/reducers/room-slice.test.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ import roomReducer, {
 
 const initialState: RoomState = {};
 
-describe('atm slice reducer', () => {
```
Comment (jim-b2c2) [inline src/reducers/room-slice.test.ts:13]: good spot

File: src/reducers/trades-slice.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  data: any;
+  error?: string;
+  filter?: string;
+};
+
+type TradeDate = {
+  value: string;
+}
+
+type Quantity = {
```
Comment (jim-b2c2) [inline src/reducers/trades-slice.ts:17]: not sure if it is better to import actual types from service lib
`import { Quantity } from '../../web-grpc-client/proto/service_pb';` Feel free to ignore if you have tried and it didnt seem to work out

File: src/utils/trade/api.ts [hunk 4]
```diff
@@ -279,5 +285,30 @@ export const createClient = (): ClientType => {
...
+          const request = new TradePageRequest();
+
+          return new Promise((resolve) => {
+            result = client.query(request);
+
+            result.on('data', (response) => {
+              resolve(response?.toObject());
+            });
+
+            result.on('end', (status) => {
```
Comment (jim-b2c2) [inline src/utils/trade/api.ts:305]: I guess remove to we actual have a good use case for it

**Review by jim-b2c2 — APPROVED** [review]

File: src/components/trades/TradesList.tsx (hunk 1 above)
Comment (laura-b2c2) [reply src/components/trades/TradesList.tsx]: Thanks :) fixed now

File: src/reducers/trades-slice.ts (hunk 3 above)
Comment (laura-b2c2) [reply src/reducers/trades-slice.ts]: I had not tried - should be sorted now :)

File: src/components/trades/TradesList.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+`
+
+const Row = ({ style, data, index }: ListChildComponentProps<TradeData[]>) => {
+  const rowData = data[index].data;
+  const rowMetaData = data[index].metadata;
+  const notional = rowData.quantity || rowData.notional;
+  return (
+    <StyledRow style={{
+      ...style, 
+      backgroundColor: index % 2 ? 'var(--color-grey-medium)' : 'var(--color-grey-very-dark)'
```
Comment (joeB2C2) [inline src/components/trades/TradesList.tsx:50]: is there a reason this isn't using odd/even css?

File: src/components/trades/TradesList.tsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          height={height - 34}
+          itemCount={trades.length}
+          itemData={trades}
+          itemSize={35}
+          width={TABLE_WIDTH}
+          style={{
+          overflow: 'visible auto',
+          }}
+        >
+          {Row}
```
Comment (joeB2C2) [inline src/components/trades/TradesList.tsx:109]: Confused why this isn't <Row />

Comment (joeB2C2) [reply src/components/trades/TradesList.tsx:109]: `<Row />`

File: src/reducers/trades-slice.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    const [query, cancel] = tradeApi.query();
+
+    thunkAPI.signal.addEventListener('abort', () => {
+      cancel();
+    });
+
+    const result = await query();
+    return result;
+  },
+);
```
Comment (joeB2C2) [inline src/reducers/trades-slice.ts:53]: simple enough isn't it! I like createAsyncThunk - no idea if it handles `.aborted` or `.rejected` fulfills that?

**Review by joeB2C2 — APPROVED** [review]
Happy for this to merge - comments more curiosity than direction

File: src/components/trades/TradesList.tsx (hunk 5 above)
Comment (laura-b2c2) [reply src/components/trades/TradesList.tsx:50]: I thought about using classes but this seemed easier at the time. I can change it if we would prefer to use classes for this.

File: src/components/trades/TradesList.tsx (hunk 6 above)
Comment (laura-b2c2) [reply src/components/trades/TradesList.tsx:109]: It's a React Window thing - we don't actually render `Row` here react window takes the Row as children and renders them with the correct data when they are in view.

File: src/components/trades/TradesList.tsx (hunk 5 above)
Comment (joeB2C2) [reply src/components/trades/TradesList.tsx:50]: was thinking this kind of thing: 
```
  &:nth-child(odd) {
    background: #efefef;
  }
```

Comment (laura-b2c2) [reply src/components/trades/TradesList.tsx:50]: ahh - can't fo that due to virtualisation :)

Comment (joeB2C2) [reply src/components/trades/TradesList.tsx:50]: got it 👍

**Pull request merged**